### PR TITLE
Move `SemaphoreStep` states into a single map to try to simplify the implementation

### PR DESCRIPTION
**Pull request by dwnusbaum**
See https://github.com/jenkinsci/workflow-support-plugin/pull/258#issuecomment-2026171643.

Personally, I feel like this is still pretty comparable to https://github.com/jenkinsci/workflow-support-plugin/pull/258 in terms of complexity. It seems a little better in terms of synchronization complexity, but the state transition complexity seems about the same or worse. The issue is that the four maps did not represent distinct non-overlapping states and so it is still not that clear what is going on in the new code. If we are purely going for simplicity, something like https://github.com/jenkinsci/workflow-support-plugin/pull/259 would be my vote, then #258, then this PR.

If we want to go with this approach, I think this PR would need a PCT run because it seems possible that is accidentally broke some typical use case.

### CODE REVIEW

File: src/test/java/org/jenkinsci/plugins/workflow/test/steps/SemaphoreStep.java [hunk 1]
```diff
@@ -134,12 +133,16 @@ public static void failure(String k, Throwable error) {
...
+            Object keyState = s.keyStates.get(k);
+            if (keyState == null) {
                 LOGGER.info(() -> "Planning to unblock " + k + " as failure");
-                s.errors.put(k, error);
+                s.keyStates.put(k, new ImmediateFailureState(error));
                 return;
+            } else if (keyState instanceof StartedState) {
+                c = getContext(s, k);
+            } else {
+                throw new IllegalStateException("Cannot mark " + k + " as failure from state " + keyState);
```
Comment (dwnusbaum) [inline src/test/java/org/jenkinsci/plugins/workflow/test/steps/SemaphoreStep.java:144]: ditto

File: src/test/java/org/jenkinsci/plugins/workflow/test/steps/SemaphoreStep.java [hunk 2]
```diff
@@ -222,7 +228,7 @@ public static class Execution extends AbstractStepExecutionImpl {
         @Override public void stop(Throwable cause) throws Exception {
             State s = State.get();
             synchronized (s) {
-                s.contexts.remove(k);
+                s.keyStates.remove(k);
```
Comment (dwnusbaum) [inline src/test/java/org/jenkinsci/plugins/workflow/test/steps/SemaphoreStep.java:231]: I am not sure that this behavior is the same. Would have to think about it some more.

File: src/test/java/org/jenkinsci/plugins/workflow/test/steps/SemaphoreStep.java [hunk 3]
```diff
@@ -231,7 +237,7 @@ public static class Execution extends AbstractStepExecutionImpl {
         @Override public String getStatus() {
             State s = State.get();
             synchronized (s) {
-                return s.contexts.containsKey(k) ? "waiting on " + k : "finished " + k;
+                return s.keyStates.get(k) == null ? "waiting on " + k : "finished " + k ;
```
Comment (dwnusbaum) [inline src/test/java/org/jenkinsci/plugins/workflow/test/steps/SemaphoreStep.java:240]: I think this behavior is the same as before, but I'm not 100% sure. Notably, it returns "waiting on" after `success` and `failure` have been called (the old code did too). To fix that, I think we'd need to add another state.

File: src/test/java/org/jenkinsci/plugins/workflow/test/steps/SemaphoreStep.java [hunk 4]
```diff
@@ -206,14 +209,11 @@ public static class Execution extends AbstractStepExecutionImpl {
...
                 LOGGER.info(() -> "Blocking " + k);
-                String c = Jenkins.XSTREAM.toXML(getContext());
-                synchronized (s) {
-                    s.contexts.put(k, c);
-                }
                 sync = false;
             }
             synchronized (s) {
-                s.started.add(k);
+                // Even if we completed immediately, set the state to StartedState so waitForSuccess knows to stop waiting.
```
Comment (jglick) [inline src/test/java/org/jenkinsci/plugins/workflow/test/steps/SemaphoreStep.java:215]: Why? I would expect the wait to just wait for the key to be present, with any value.

Also did you mean
```suggestion
                // Even if we completed immediately, set the state to StartedState so waitForStart knows to stop waiting.
```
?

File: src/test/java/org/jenkinsci/plugins/workflow/test/steps/SemaphoreStep.java (hunk 2 above)
Comment (jglick) [reply src/test/java/org/jenkinsci/plugins/workflow/test/steps/SemaphoreStep.java:225]: Seems safer to switch to the failure state. Should be idempotent right? There is no real reason to remove entries from the map.

File: src/test/java/org/jenkinsci/plugins/workflow/test/steps/SemaphoreStep.java (hunk 3 above)
Comment (jglick) [reply src/test/java/org/jenkinsci/plugins/workflow/test/steps/SemaphoreStep.java:234]: Does not look right. As above, just do not remove entries from the map. Then this will say it is finished when it is.

File: src/test/java/org/jenkinsci/plugins/workflow/test/steps/SemaphoreStep.java [hunk 5]
```diff
@@ -255,5 +255,38 @@ public static class Execution extends AbstractStepExecutionImpl {
 
     }
 
+    // Marker interface just for clarity.
+    private interface KeyState { }
+
+    /**
+     * The step has not yet started, but will succeed immediately when it does start.
+     */
+    private static class ImmediateSuccessState implements KeyState {
```
Comment (jglick) [inline src/test/java/org/jenkinsci/plugins/workflow/test/steps/SemaphoreStep.java:264]: Clearer IMO to just call it a success state. The timing of that success is a matter for the code moving it around in the map.

File: src/test/java/org/jenkinsci/plugins/workflow/test/steps/SemaphoreStep.java [hunk 6]
```diff
@@ -255,5 +255,38 @@ public static class Execution extends AbstractStepExecutionImpl {
...
+     */
+    private static class ImmediateFailureState implements KeyState {
+        private final Throwable error;
+        private ImmediateFailureState(Throwable error) {
+            this.error = error;
+        }
+    }
+
+    /**
+     * The step has at least started, and may have finished.
```
Comment (jglick) [inline src/test/java/org/jenkinsci/plugins/workflow/test/steps/SemaphoreStep.java:282]: > may have finished

??

**Review by jglick — COMMENTED** [review]
Not sure the details of the states are right, but the overall structure seems clearer to me.

**Comment by jglick** [comment]
> this PR would need a PCT run

We should do a PCT run for any proposed change to this class. (`jenkinsci/bom` and/or @cloudbees internal)

File: src/test/java/org/jenkinsci/plugins/workflow/test/steps/SemaphoreStep.java (hunk 4 above)
Comment (dwnusbaum) [reply src/test/java/org/jenkinsci/plugins/workflow/test/steps/SemaphoreStep.java:215]: Because otherwise if you call `SemaphoreStep.success` or `SemaphoreStep.failure` and then `SemaphoreStep.waitForStart` _without even running a `Pipeline` that uses the `semaphore` step_, `waitForStart` will complete immediately.

Comment (dwnusbaum) [reply src/test/java/org/jenkinsci/plugins/workflow/test/steps/SemaphoreStep.java:215]: Basically, to really get this right, there needs to be at least 5 states, and we need to add a few more transitions that currently do not exist:
* No entry in the map 
* ImmediateFailure (the step will fail, if/when it starts)
* ImmediateSuccess (the step will succeed, if/when it starts)
* Started/Waiting/Blocked (the step has started, and has not yet completed)
* Finished (the step has finished, with some result) (does not currently exist, we'd need to add various transitions to this state)

And I do not think that we can introduce the `Finished` state without also introducing various behavior changes in edge cases that might affect existing tests. We'd definitely need a PCT run.

File: src/test/java/org/jenkinsci/plugins/workflow/test/steps/SemaphoreStep.java (hunk 3 above)
Comment (dwnusbaum) [reply src/test/java/org/jenkinsci/plugins/workflow/test/steps/SemaphoreStep.java]: Adjusted in [d05eafd](https://github.com/jenkinsci/workflow-support-plugin/pull/260/commits/d05eafd27820d48db69890d04658053da4887743).

File: src/test/java/org/jenkinsci/plugins/workflow/test/steps/SemaphoreStep.java (hunk 2 above)
Comment (dwnusbaum) [reply src/test/java/org/jenkinsci/plugins/workflow/test/steps/SemaphoreStep.java]: Adjusted in [d05eafd](https://github.com/jenkinsci/workflow-support-plugin/pull/260/commits/d05eafd27820d48db69890d04658053da4887743)

File: src/test/java/org/jenkinsci/plugins/workflow/test/steps/SemaphoreStep.java (hunk 5 above)
Comment (dwnusbaum) [reply src/test/java/org/jenkinsci/plugins/workflow/test/steps/SemaphoreStep.java]: I think it is important to be able to distinguish "the step will succeed, when/if it starts" from "the step ran and finished". Maybe not for the common case, but for edge cases it seems important, and the old code was able to distinguish these states. Either way, I adjusted things in [d05eafd](https://github.com/jenkinsci/workflow-support-plugin/pull/260/commits/d05eafd27820d48db69890d04658053da4887743).

File: src/test/java/org/jenkinsci/plugins/workflow/test/steps/SemaphoreStep.java [hunk 7]
```diff
@@ -108,12 +108,15 @@ public static void success(String k, Object returnValue) {
         StepContext c;
         synchronized (s) {
             KeyState keyState = s.keyStates.get(k);
-            if (!(keyState instanceof StartedState)) {
+            if (keyState == null || keyState instanceof WillSucceedState || keyState instanceof WillFailState) {
```
Comment (jglick) [inline src/test/java/org/jenkinsci/plugins/workflow/test/steps/SemaphoreStep.java:111]: Maybe introduce a common supertype `WillFinishState`.

**Review by jglick — APPROVED** [review]

File: src/test/java/org/jenkinsci/plugins/workflow/test/steps/SemaphoreStep.java [hunk 8]
```diff
@@ -112,12 +107,16 @@ public static void success(String k, Object returnValue) {
...
         synchronized (s) {
-            if (!s.contexts.containsKey(k)) {
+            KeyState keyState = s.keyStates.get(k);
+            if (keyState == null || keyState instanceof WillSucceedState || keyState instanceof WillFailState) {
                 LOGGER.info(() -> "Planning to unblock " + k + " as success");
-                s.returnValues.put(k, returnValue);
+                s.keyStates.put(k, new WillSucceedState(returnValue));
                 return;
+            } else if (keyState instanceof FinishedState) {
+                throw new IllegalStateException(k + " already finished");
```
Comment (dwnusbaum) [inline src/test/java/org/jenkinsci/plugins/workflow/test/steps/SemaphoreStep.java:116]: This breaks tests. In particular at least `workflow-step-api / org.jenkinsci.plugins.workflow.steps.GeneralNonBlockingStepExecutionTest.stop` and `workflow-cps / org.jenkinsci.plugins.workflow.cps.CpsFlowExecutionTest.stepsAreStoppedWhenCpsVmExecutorServiceHandlesUncaughtException`. I'll take a look.

Comment (dwnusbaum) [reply src/test/java/org/jenkinsci/plugins/workflow/test/steps/SemaphoreStep.java:116]: `GeneralNonBlockingStepExecutionTest.stop` I think just has a redundant call to `SemaphoreStep.success` that could be removed.

`CpsFlowExecutionTest.stepsAreStoppedWhenCpsVmExecutorServiceHandlesUncaughtException` is a little unique, basically the test was using `semaphore` as a way to call `StepContext.onSuccess` after a Pipeline died. To preserve that test's behavior without changes, we'd have to make `FinishedState` keep the serialized context around. Probably clearer to update that test to use a custom step instead.

**Comment by dwnusbaum** [comment]
I have no plans to finish this up, so I am closing it; there are two tests mentioned above in other plugins would require changes either here or to the tests depending on the desired behavior.
